fdbserver/workloads/DataDistributionMetrics.actor.cpp

```diff
@@ -26,52 +26,119 @@
+				// In addition, the first key in the result will be the last one <= startKey (Condition #1)
+				state KeySelector begin =
+				    KeySelectorRef(startKey.withPrefix(ddStatsRange.begin, startKey.arena()), true, 0);
+				// Find the last key less than endKey, move forward 2 keys, and use this key as the (exclusive) end of
+				// the range. If we didn't read through the end of the range, then the second last key
+				// in the result will be the last key less than endKey. (Condition #2)
+				state KeySelector end = KeySelectorRef(endKey.withPrefix(ddStatsRange.begin, endKey.arena()), false, 2);
+				Standalone<RangeResultRef> result =
+				    wait(tr->getRange(begin, end, GetRangeLimits(CLIENT_KNOBS->SHARD_COUNT_LIMIT)));
+				// Condition #1 and #2 can be broken if multiple rpc calls happened in one getRange
+				if (result.size() > 1) {
+					if (result[0].key <= begin.getKey() && result[1].key > begin.getKey()) {
+						TraceEvent(SevDebug, "DDMetricsConsistencyTest")
+						    .detail("Size", result.size())
+						    .detail("FirstKey", result[0].key.toString())
+						    .detail("SecondKey", result[1].key.toString())
+						    .detail("BeginKeySelector", begin.toString());
+					} else {
+						++self->errors;
+						TraceEvent(SevError, "TestFailure")
+						    .detail("Reason", "Result mismatches the given begin selector")
+						    .detail("Size", result.size())
+						    .detail("FirstKey", result[0].key.toString())
+						    .detail("SecondKey", result[1].key.toString())
+						    .detail("BeginKeySelector", begin.toString());
+					}
+					if (result[result.size() - 1].key >= end.getKey() && result[result.size() - 2].key < end.getKey()) {
+						TraceEvent(SevDebug, "DDMetricsConsistencyTest")
+						    .detail("Size", result.size())
+						    .detail("LastKey", result[result.size() - 1].key.toString())
+						    .detail("SecondLastKey", result[result.size() - 2].key.toString())
+						    .detail("EndKeySelector", end.toString());
+					} else {
+						++self->errors;
+						TraceEvent(SevError, "TestFailure")
+						    .detail("Reason", "Result mismatches the given end selector")
+						    .detail("Size", result.size())
+						    .detail("FirstKey", result[result.size() - 1].key.toString())
+						    .detail("SecondKey", result[result.size() - 2].key.toString())
+						    .detail("EndKeySelector", end.toString());
 					}
-					wait(tr.commit());
-					++self->transactionsCommitted;
-					break;
-				} catch (Error& e) {
-					wait(tr.onError(e));
 				}
+			} catch (Error& e) {
+				// Ignore timed_out error and cross_module_read, the end key selector may read through the end
+				if (e.code() == error_code_timed_out || e.code() == error_code_special_keys_cross_module_read) continue;
+				TraceEvent(SevDebug, "FailedToRetrieveDDMetrics").detail("Error", e.what());
```

Comment:
.error() can provide the error message and error code and stack trace. :)

Reply:
Yeah, I should use this. Thanks a lot.